src/Storages/RabbitMQ/ReadBufferFromRabbitMQConsumer.cpp

```diff
@@ -0,0 +1,268 @@
+     * then for uniformity this parameter should always be set as fanout-exchange type. In current implementation, the exchange,
+     * to which messages a published, will be bound to the exchange of the needed type, which will distribute messages according to its type.
+     */
+    consumer_channel->declareExchange(exchange_name, AMQP::fanout).onError([&](const char * message)
+    {
+        exchange_declared = false;
+        LOG_ERROR(log, "Failed to declare fanout exchange: " << message);
+    });
+
+    if (hash_exchange)
+    {
+        current_exchange_name = exchange_name + "_hash";
+        consumer_channel->declareExchange(current_exchange_name, AMQP::consistent_hash).onError([&](const char * message)
+        {
+            exchange_declared = false;
+        });
+
+        consumer_channel->bindExchange(exchange_name, current_exchange_name, routing_key).onError([&](const char * message)
+        {
+            exchange_declared = false;
+        });
+    }
+    else
+    {
+        current_exchange_name = exchange_name + "_direct";
+        consumer_channel->declareExchange(current_exchange_name, AMQP::direct).onError([&](const char * message)
+        {
+            exchange_declared = false;
+        });
+
+        consumer_channel->bindExchange(exchange_name, current_exchange_name, routing_key).onError([&](const char * message)
+        {
+            exchange_declared = false;
+        });
+    }
+}
+
+
+void ReadBufferFromRabbitMQConsumer::initQueueBindings(const size_t queue_id)
+{
+    if (!exchange_declared)
+    {
+        initExchange();
+        exchange_declared = true;
+    }
+
+    bool bindings_ok = false, bindings_error = false;
+
+    consumer_channel->declareQueue(AMQP::exclusive)
+    .onSuccess([&](const std::string &  queue_name_, int /* msgcount */, int /* consumercount */)
+    {
+        queues.emplace_back(queue_name_);
+
+        String binding_key = routing_key;
+
+        if (bind_by_id && !hash_exchange)
+        {
+            if (queues.size() == 1)
+            {
+                binding_key = routing_key + "_" + std::to_string(channel_id);
+            }
+            else
+            {
+                binding_key = routing_key + "_" + std::to_string(channel_id + queue_id);
+            }
+        }
+
+        LOG_TRACE(log, "Queue " + queue_name_ + " is bound by key " + binding_key);
+
+        consumer_channel->bindQueue(current_exchange_name, queue_name_, binding_key)
+        .onSuccess([&]
+        {
+            bindings_ok = true;
+        })
+        .onError([&](const char * message)
+        {
+            bindings_error = true;
+            LOG_ERROR(log, "Failed to create queue binding: " << message);
+        });
+    })
+    .onError([&](const char * message)
+    {
+        bindings_error = true;
+        LOG_ERROR(log, "Failed to declare queue on the channel: " << message);
+    });
+
+    while (!bindings_ok && !bindings_error)
```

Comment:
These variables can be updated in a separate thread? Need a comment about, now it seems like endless loop :)